### ISSUE TITLE
No report generated when using multiple JBehave threads

### ISSUE THREAD

**Issue by csabakoncz** [serenity-bdd/serenity-jbehave#89]
I prepared a sample project here: 
https://github.com/csabakoncz/serenity-projects/tree/master/serenity-timeouts
(originally to test the recently introduced per-story timeout feature).

Running the tests with
`mvn clean verify -Djbehave.threads=1`
produces the beautiful Serenity reports.

However
`mvn clean verify -Djbehave.threads=2`
results in an empty report.

The tests run rather slowly, because they test timeouts, however, the report is missing even if the steps implementation is removed (i.e. the steps are reported as pending).

I am using serenity-jbehave 1.12.0

**Comment by wakaleo** [comment]
Thanks, we will investigate this further - the sample project will make things a lot easier to troubleshoot.

**Comment by csabakoncz** [comment]
Hi,
Could you reproduce the issue? Or it is something that shows up only at me?
Thanks,
Csaba

**Comment by csabakoncz** [comment]
I think this is rather an issue with JBehave and not with Serenity. When using multiple threads `SerenityReporter.afterStory()` is not invoked for 'AfterStories'.

This is what I see in JBehave 4.0.5 `PerformableTree`:

```
public void performBeforeOrAfterStories(RunContext context, Stage stage) {
        String storyPath = StringUtils.capitalize(stage.name().toLowerCase()) + "Stories";
        context.currentPath(storyPath);
        context.reporter().beforeStory(new Story(storyPath), false);
        try {
            (stage == Stage.BEFORE ? root.beforeSteps : root.afterSteps).perform(context);
        } catch (InterruptedException e) {
            throw new UUIDExceptionWrapper(e);
        } finally {
            if (context.reporter() instanceof ConcurrentStoryReporter) {
                ((ConcurrentStoryReporter) context.reporter()).invokeDelayed();
            }
        }
        context.reporter().afterStory(false);
    }
```

`ConcurrentStoryReporter.invokeDelayed()` is only invoked after calling `beforeStory()`. As `AfterStories` is the last  one its `afterStory()` is never broadcast.

**Comment by DamianGuzier** [comment]
@csabakoncz, do you find solution?
@wakaleo do you investigate this problem?

We have got the same problem.
Serenity-Core version 1.2.2 (version does not matter. We trying 1.1.42, 1.2.4, 1.2.5.rc-X)
Serenity-Jbehave version 1.21.0 (in 1.22.0, 1.23.0 and 1.24.0 stories are not running in parallel mode with .useThreads(2). They are running sequentially)
JBehave version 4.1
We run tests as below
```
public TestRunnerUI() {
        findStoriesIn("/stories/selenium");
        Embedder embedder = configuredEmbedder();
        embedder
            .embedderControls()
            .useThreads(2);
        configuration()
            .usePendingStepStrategy(new FailingUponPendingStep())
            .storyReporterBuilder().withMultiThreading( true );
        configuration()
            .storyReporterBuilder().withReporters(new SerenityReporter(getSystemConfiguration()));
    }
```
Story file as below:
<img width="295" alt="zrzut ekranu 2017-04-28 o 09 27 26" src="https://cloud.githubusercontent.com/assets/25481798/25518685/0c6643c2-2bf5-11e7-8039-f01feeb9b18f.png">

When we configure  **useThreads(2)** (or more threads) stories are running in parallel mode fine but results in an empty report. Overral Test result is empty, Requirements Overview display as unTestes:
<img width="1058" alt="zrzut ekranu 2017-04-28 o 09 34 43" src="https://cloud.githubusercontent.com/assets/25481798/25518928/0649d232-2bf6-11e7-927f-364eff0c1f22.png">
<img width="1052" alt="zrzut ekranu 2017-04-28 o 09 34 54" src="https://cloud.githubusercontent.com/assets/25481798/25518930/088da42e-2bf6-11e7-8c1b-0febe7d08a52.png">

When we configure  **useThreads(1)** report is fine
If I can give you some details please tell me.
Are we missing some configuration? Or maybe we should use some special dependency version?

Expected results is to have proper report generated both in .useThreads(1) and .useThreads(X)

**Comment by wakaleo** [comment]
This appears to be a JBehave bug, as mentioned above.  The afterStory() method is not being called by JBehave, so Serenity cannot generate the reports.

**Comment by DamianGuzier** [comment]
Could you propose some fix for that? What and where we should change in JBehave? We can try yours JBehave fix in our project

**Comment by wakaleo** [comment]
I have no idea, you would have to investigate the JBehave code, which I am not familiar with.

**Comment by csabakoncz** [comment]
@DamianGuzier 

I reported the issue to the JBehave team a while ago: https://jbehave.atlassian.net/browse/JBEHAVE-1174 Unfortunately received no reaction at all.

Actually, this is not the only problem. Even if you fix it, you run into the next one: when multithreaded, JBehave notifies the story listeners in a batched way: it collects the events and sends them out only after the story is finished.  This is seemingly done  in order to avoid concurrency issues when multiple threads are accessing the same story listener instance. 
Sadly this behavior does not suit Serenity: in order to correctly initialize and clean up the step factories, for example, the before and after notifications have to be delivered when they really happen.

I ended up patching JBehave classes to get the required behavior for Serenity. Then, the next thing:  SerenityReporter is not thread-safe - you have to add a number of thread local variables to make it able to keep track of stories executing in parallel.

So having Serenity+JBehave running on multiple threads requires quite some effort. 
I doubt that anyone else is doing that. 
Instead, parallel execution is achieved using the forking feature of the Maven Failsafe plugin - which is less elegant, hard to get the work evenly divided between JVM instances, but is simpler in JBehave terms.

**Comment by valfirst** [comment]
@csabakoncz,

> when multithreaded, JBehave notifies the story listeners in a batched way: it collects the events and sends them out only after the story is finished. This is seemingly done in order to avoid concurrency issues when multiple threads are accessing the same story listener instance.
Sadly this behavior does not suit Serenity: in order to correctly initialize and clean up the step factories, for example, the before and after notifications have to be delivered when they really happen.

you can change this behavior for your reporters by overriding `org.jbehave.core.reporters.StoryReporterBuilder.build(String)`, for example:

```java
new StoryReporterBuilder() {
    @Override
    public StoryReporter build(String storyPath) {
        return new DelegatingStoryReporter(super.build(storyPath), new SerenityReporter(systemConfiguration));
    }
}.with...;
```

**Comment by wakaleo** [comment]
@valfirst's response may be a work-around. If not, this is not a feature we will be supporting in Serenity natively any time soon.
